src/poly/dcrt/matrix/generic_matrix.rs

```diff
@@ -368,26 +431,46 @@ impl<T: MatrixElem> GenericMatrix<T> {
-        let mut new_matrix = Self::new_empty(&self.params, self.nrow, updated_ncol);
-        let self_f = |row_offsets: Range<usize>, col_offsets: Range<usize>| -> Vec<Vec<T>> {
-            self.block_entries(row_offsets, col_offsets)
-        };
-        new_matrix.replace_entries(0..self.nrow, 0..self.ncol, self_f);
-        debug_mem("self replaced in concat_columns");
-
-        let mut col_acc = self.ncol;
-        for (idx, other) in others.iter().enumerate() {
-            let other_f = |row_offsets: Range<usize>, col_offsets: Range<usize>| -> Vec<Vec<T>> {
-                let col_offsets = col_offsets.start - col_acc..col_offsets.end - col_acc;
-                other.block_entries(row_offsets, col_offsets)
+        #[cfg(feature = "disk")]
+        {
+            let updated_ncol = others.iter().fold(self.ncol, |acc, other| acc + other.ncol);
+            let mut new_matrix = Self::new_empty(&self.params, self.nrow, updated_ncol);
+            let self_f = |row_offsets: Range<usize>, col_offsets: Range<usize>| -> Vec<Vec<T>> {
+                self.block_entries(row_offsets, col_offsets)
             };
-            new_matrix.replace_entries(0..self.nrow, col_acc..col_acc + other.ncol, other_f);
-            debug_mem(format!("the {}-th other replaced in concat_columns", idx));
-            col_acc += other.ncol;
+            new_matrix.replace_entries(0..self.nrow, 0..self.ncol, self_f);
+            debug_mem("self replaced in concat_columns");
+
+            let mut col_acc = self.ncol;
+            for (idx, other) in others.iter().enumerate() {
+                let other_f =
+                    |row_offsets: Range<usize>, col_offsets: Range<usize>| -> Vec<Vec<T>> {
+                        let col_offsets = col_offsets.start - col_acc..col_offsets.end - col_acc;
+                        other.block_entries(row_offsets, col_offsets)
+                    };
+                new_matrix.replace_entries(0..self.nrow, col_acc..col_acc + other.ncol, other_f);
+                debug_mem(format!("the {}-th other replaced in concat_columns", idx));
+                col_acc += other.ncol;
+            }
+            debug_assert_eq!(col_acc, updated_ncol);
+            new_matrix
+        }
+        #[cfg(not(feature = "disk"))]
+        {
+            let ncol = self.ncol + others.iter().map(|x| x.ncol).sum::<usize>();
+            let result: Vec<Vec<T>> = parallel_iter!(0..self.nrow)
+                .map(|i| {
+                    let mut row: Vec<&T> = Vec::with_capacity(ncol);
+                    row.extend(self.inner[i].iter());
+                    for other in others {
+                        row.extend(other.inner[i].iter());
+                    }
+                    row.into_iter().cloned().collect()
+                })
+                .collect();
+
+            Self { inner: result, params: self.params.clone(), nrow: self.nrow, ncol }
```

Comment:
How is the benchmark different when the same `replace_entries` logic is used when the feature is not disk?

Reply:
haven't got exact number but this logic removes unnecessary `new_empty` `block_entries` `replace_entries` which desn't need to be exist in memory (vector) based implementation so expect optimal